eng/scripts/Language-Settings.ps1

```diff
@@ -10,54 +10,187 @@ $PackageRepositoryUri = "https://repo1.maven.org/maven2"
-      return $null
+    if ($YmlFileSdkType -ne $SdkType) {
+      Write-Host "SdkType in yml file is '$YmlFileSdkType' which is not '$SdkType', skipping..."
+      continue
     }
-
-    if ($projectData.project.psobject.properties.name -notcontains "groupid" -or !$projectData.project.groupId) {
-      Write-Host "$projectPath doesn't have a defined groupId so skipping this pom."
-      return $null
+    # ServiceDirectory
+    $serviceDirFromYml = GetValueSafelyFrom-Yaml $ymlFileContent @("extends", "parameters", "ServiceDirectory")
+    if (-not $serviceDirFromYml) {
+      # Log the error and skip this yml file here if there's no ServiceDirectory entry
+      LogWarning "$ymlFile does not have a ServiceDirectory entry, skipping..."
+      continue
     }
-
-    $projectPkgName = $projectData.project.artifactId
-    $pkgVersion = $projectData.project.version
-    $pkgGroup = $projectData.project.groupId
-
-    $pkgProp = [PackageProps]::new($projectPkgName, $pkgVersion.ToString(), $pkgPath, $serviceDirectory, $pkgGroup)
-    if ($projectPkgName -match "mgmt" -or $projectPkgName -match "resourcemanager")
-    {
-      $pkgProp.SdkType = "mgmt"
-    }
-    elseif ($projectPkgName -match "spring")
-    {
-      $pkgProp.SdkType = "spring"
+    else {
+      # If the serviceDirectory parameter was passed in, ensure that the ServiceDirectory
+      # entry in the ci*.yml file matches the one passed in.
+      if ($serviceDirectory) {
+        if ($serviceDirectory -ne $serviceDirFromYml) {
+          LogWarning "$ymlFile's ServiceDirectory entry does not match the serviceDirectory parameter: '$serviceDirectory'"
+        }
+      }
+      # Check whether or not the yml's serviceDirectory matches the actual path of
+      # the yml file relative to the sdkRoot
+      # Note: Need to strip off the directory seperator character which is based upon the OS
+      $computedServiceDirectory = $ymlDir.Replace($sdkRoot + [System.IO.Path]::DirectorySeparatorChar, "")
+      # .Replace and -replace have different behaviors. .Replace will not replace a backslash meaning
+      # that "foo\bar".Replace("\\","/") would result in foo\bar instead of foo/bar which is why
+      # -replace needs to be used
+      $computedServiceDirectory = $computedServiceDirectory -replace "\\", "/"
+      if ($serviceDirFromYml -ne $computedServiceDirectory) {
+        LogWarning "$ymlFile error: ServiceDirectory in the yml file, '$serviceDirFromYml' doesn't match the path relative from the sdkRoot '$computedServiceDirectory'"
+      }
     }
-    else
+    # At this point the SdkType is correct.
+    # 1. Create a hash set from the list of artifact for this service directory, for
+    #    the SdkType. This is done to ensure that only PackageInfo files are created
+    #    for the correct set of Artifacts
+    # 2. Save off AdditionalModules to add to the AdditionalValidationPackages entry
+    $ArtifactsHashSet = New-Object 'System.Collections.Generic.HashSet[String]'
```

Comment:
```suggestion
    $ArtifactsHashSet = @()
```
nit: You can initialize this directly in PS

Reply:
@weshaggard I'm intentionally using a HashSet and the suggested change would change it to a Hashtable. I don't have key/value pairs, just a unique collection values and this is the right data structure for what I need.